src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,890 @@
+    const int asid = 0;
+    const Addr pc = 0; // TODO extract this from inflightinst, but otherwise
+                       //      only used for tracing/debugging.
+    const MasterID mid = sdCPUThread._cpuPtr->dataMasterId();
+    ThreadContext* const tc = sdCPUThread.getThreadContext();
+    const ContextID cid = tc->contextId();
+
+    RequestPtr req = make_shared<Request>(asid, addr, size, flags, mid, pc,
+                                          cid);
+
+    // For tracing?
+    req->taskId(sdCPUThread._cpuPtr->taskId());
+
+    const unsigned cache_line_size = sdCPUThread._cpuPtr->cacheLineSize();
+    const Addr split_addr = roundDown(addr + size - 1, cache_line_size);
+
+    weak_ptr<InflightInst> inst = inst_ptr;
+
+    if (split_addr > addr) { // The request spans two cache lines
+        assert(!(req->isLLSC() || req->isMmappedIpr())); // don't deal with
+                                                         // weird cases for now
+
+        RequestPtr req1, req2;
+
+        req->splitOnVaddr(split_addr, req1, req2);
+
+        shared_ptr<SplitRequest> split_acc = make_shared<SplitRequest>();
+        split_acc->main = req;
+
+
+        auto callback1 = [this, inst, split_acc](Fault fault,
+                                                 const RequestPtr& req) {
+            sdCPUThread.onDataAddrTranslated(inst, fault, req, false,
+                                             nullptr, split_acc, false);
+        };
+        auto callback2 = [this, inst, split_acc](Fault fault,
+                                                 const RequestPtr& req) {
+            sdCPUThread.onDataAddrTranslated(inst, fault, req, false,
+                                             nullptr, split_acc, true);
+        };
+
+        if (!sdCPUThread._cpuPtr->requestDataAddrTranslation(req1, tc,
+                                                             callback1)) {
+            panic("The CPU rejected my dtb request and I haven't been"
+                  " programmed to know how to continue!!!");
+            // TODO request later
+        }
+
+        if (!sdCPUThread._cpuPtr->requestDataAddrTranslation(req2, tc,
+                                                             callback2)) {
+            panic("The CPU rejected my dtb request and I haven't been"
+                  " programmed to know how to continue!!!");
+            // TODO request later
+        }
+
+        return NoFault;
+    }
+
+
+    auto callback = [this, inst](Fault fault, const RequestPtr& req) {
+        sdCPUThread.onDataAddrTranslated(inst, fault, req, false,
+                                         nullptr);
+    };
+
+    if (!sdCPUThread._cpuPtr->requestDataAddrTranslation(req, tc, callback)) {
+        panic("The CPU rejected my dtb request and I haven't been programmed"
+              " to know how to continue!!!");
+        // TODO request later
+    }
+
+    return NoFault;
+}
+
+Fault
+SDCPUThread::MemIface::writeMem(shared_ptr<InflightInst> inst_ptr,
```

Comment:
There's a lot of  duplicated code between read and write. (This is true in most of the functions that relate to these two things, though this is the worst so far.) Is there any way to factor some of this out?